### ISSUE TITLE
Find out how to send files larger than 150mb from watch to smartphone

### ISSUE THREAD

**Issue by danmida** [DanielBlomberg1/mimir#12]
At the time of testing, somewhere between 150mb to 200mb is a current limit to sending a file from watch to smartphone before emulator watch crashes. 

Error log states watch device is out of memory.

Customer stated that log data could potentially be in hundreds of megabytes. Potentially, they could work around the limitation of file size.

Compress the log on watch and decompress on phone?

My original thought is to chunk the files.

**Comment by danmida** [comment]
E  FATAL EXCEPTION: main
                                                                                                    Process: com.mobilewizards.logging_app, PID: 2954
                                                                                                    java.lang.OutOfMemoryError: Failed to allocate a 308267568 byte allocation with 4054014 free bytes and 188MB until OOM, target footprint 8108030, growth limit 201326592
                                                                                                    	at kotlin.io.FilesKt__FileReadWriteKt.readBytes(FileReadWrite.kt:68)
                                                                                                    	at com.mobilewizards.logging_app.SendSurveysActivity.sendCsvFileToPhone(SendSurveysActivity.kt:380)
                                                                                                    	at com.mobilewizards.logging_app.SendSurveysActivity.access$sendCsvFileToPhone(SendSurveysActivity.kt:42)
                                                                                                    	at com.mobilewizards.logging_app.SendSurveysActivity$sendFiles$1.invoke(SendSurveysActivity.kt:312)
                                                                                                    	at com.mobilewizards.logging_app.SendSurveysActivity$sendFiles$1.invoke(SendSurveysActivity.kt:298)
                                                                                                    	at com.mobilewizards.logging_app.SendSurveysActivity$getPhoneNodeId$1.invoke(SendSurveysActivity.kt:349)
                                                                                                    	at com.mobilewizards.logging_app.SendSurveysActivity$getPhoneNodeId$1.invoke(SendSurveysActivity.kt:344)
                                                                                                    	at com.mobilewizards.logging_app.SendSurveysActivity.getPhoneNodeId$lambda$15(SendSurveysActivity.kt:344)
                                                                                                    	at com.mobilewizards.logging_app.SendSurveysActivity.$r8$lambda$lUqwBCdDLa-W3NRyoQd5v12tKwk(Unknown Source:0)
                                                                                                    	at com.mobilewizards.logging_app.SendSurveysActivity$$ExternalSyntheticLambda0.onSuccess(D8$$SyntheticClass:0)
                                                                                                    	at com.google.android.gms.tasks.zzm.run(com.google.android.gms:play-services-tasks@@18.0.1:1)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:942)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7898)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)

**Comment by danmida** [comment]
Was able to optimize checksum so that instead of being able to send 150mb files on emulator watch, I was able to send 300mb file before running into device out of memory error with larger files.

This limit may be lower on real physical watches. 

Anything higher than the limit requires editing transfer protocol. for example sending in chunks, compressing data etc.

**Comment by danmida** [comment]
I tested the first commit that had working file transfer to test out if checksum function have any more memory impact.

I managed to send 300mb file with success. 400mb file failed to go through, with same out of memory error.

**Comment by danmida** [comment]
The way on how the file is processed/streamed/read/etc had a big memory impact. The whole file was loaded into memory which in turn would cause watch to crash on very large files.

I optimized the code to process the file as chunks and sent to phone. It does not seem that any code changes was required for phone side fortunately.

I tested sending with 1 gigabyte file and it seems to work. On one random case, the smartphone app crashed on me but it seemed to be a rare occurence as it happened only once.
